src/Compilers/Core/Portable/Compilation/Expression.cs

```diff
@@ -0,0 +1,380 @@
+            }
+        }
+    }
+
+    public class CompoundAssignment : IExpressionStatement
+    {
+        CompoundAssignmentExpression compoundAssignment;
+
+        public CompoundAssignment(IReference target, IExpression value, BinaryOperationKind binaryKind, IMethodSymbol operatorMethod, SyntaxNode syntax)
+        {
+            this.compoundAssignment = new CompoundAssignmentExpression(target, value, binaryKind, operatorMethod, syntax);
+            this.Syntax = syntax;
+        }
+
+        public SyntaxNode Syntax { get; set; }
+
+        public OperationKind Kind
+        {
+            get { return OperationKind.ExpressionStatement; }
+        }
+
+        public IExpression Expression
+        {
+            get { return this.compoundAssignment; }
+        }
+
+        class CompoundAssignmentExpression : ICompoundAssignment
+        {
+            public CompoundAssignmentExpression(IReference target, IExpression value, BinaryOperationKind binaryKind, IMethodSymbol operatorMethod, SyntaxNode syntax)
+            {
+                this.Target = target;
+                this.Value = value;
+                this.BinaryKind = binaryKind;
+                this.Operator = operatorMethod;
+                this.Syntax = syntax;
+            }
+
+            public IReference Target { get; set; }
+
+            public IExpression Value { get; set; }
+
+            public BinaryOperationKind BinaryKind { get; set; }
+
+            public IMethodSymbol Operator { get; set; }
+
+            public SyntaxNode Syntax { get; set; }
+
+            public ITypeSymbol ResultType
+            {
+                get { return this.Target.ResultType; }
+            }
+
+            public OperationKind Kind
+            {
+                get { return OperationKind.CompoundAssignment; }
+            }
+
+            public object ConstantValue
+            {
+                get { return null; }
+            }
+
+            public bool UsesOperatorMethod
+            {
+                get { return this.Operator != null; }
+            }
+        }
+    }
+
+    public class IntegerLiteral : ILiteral
+    {
+        long value;
+
+        public IntegerLiteral(long value, ITypeSymbol resultType, SyntaxNode syntax)
+        {
+            this.value = value;
+            this.ResultType = resultType;
+            this.Syntax = syntax;
+        }
+
+        public LiteralKind LiteralClass
+        {
+            get { return LiteralKind.Integer; }
+        }
+
+        public string Spelling
+        {
+            get { return this.value.ToString(); }
+        }
+
+        public ITypeSymbol ResultType { get; set; }
+
+        public OperationKind Kind
+        {
+            get { return OperationKind.Literal; }
+        }
+
+        public object ConstantValue
+        {
+            get { return this.value; }
+        }
+
+        public SyntaxNode Syntax { get; set; }
+    }
+
+    public class Binary : IBinary
```

Comment:
Is this class used? That is, are we synthesizing operations or always using BoundExpression implementations?